src/Build/TelemetryInfra/InternalTelemetryConsumingLogger.cs

```diff
@@ -75,6 +82,22 @@ private void FlushDataIntoConsoleIfRequested()
         Console.WriteLine("==========================================");
     }
 
+    private void FlushDataIntoJsonFileIfRequested()
+    {
+        const string jsonFileNameVariable = "MSBUILDNODETELEMETRYFILENAME";
+        var jsonFilePath = Environment.GetEnvironmentVariable(jsonFileNameVariable);
+        if (string.IsNullOrEmpty(jsonFilePath))
+        {
+            return;
+        }
+
+        var telemetryTags = _workerNodeTelemetryData.AsActivityDataHolder(true, true)?.GetActivityProperties();
+
+        using var stream = File.OpenWrite(jsonFilePath);
+        stream.SetLength(0);
+        JsonSerializer.Serialize(stream, telemetryTags, new JsonSerializerOptions() { WriteIndented = true });
```

Comment:
if you specify `WriteIndented` do you need to do the same in custom deserializers?

Reply:
no, it has no effect on deserialization